README.md
# TFlite Audio Plugin for Flutter

[![pub package](https://img.shields.io/pub/v/tflite_audio.svg?label=version&color=blue)](https://pub.dev/packages/tflite_audio)
[![likes](https://badges.bar/tflite_audio/likes)](https://pub.dev/packages/tflite_audio/score)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![style: effective dart](https://img.shields.io/badge/style-effective_dart-40c4ff.svg)](https://pub.dev/packages/effective_dart)

<br>

Audio classification Tflite package for flutter (iOS & Android). Can also support Google Teachable Machine models. 

If you are a complete newbie to audio classification, you can read the tutorial [here](https://carolinamalbuquerque.medium.com/audio-recognition-using-tensorflow-lite-in-flutter-application-8a4ad39964ae). Credit to [Carolina](https://github.com/cmalbuquerque) for writing a comprehensive article.

To keep this project alive, consider giving a star or a like. Pull requests or bug reports are also welcome.

<br>

Recording            |  Inference result
:-------------------------:|:-------------------------:
![](https://github.com/Caldarie/flutter_tflite_audio/blob/master/pictures/finish.png) | ![](https://github.com/Caldarie/flutter_tflite_audio/blob/master/pictures/start.png) 

<br>

## Table of Contents

 * [About this plugin](#about-this-plugin)
 * [Known Issues/Commonly asked questions](#known-issuescommonly-asked-questions)
 * [Please read if you are using Google's Teachable Machine. Otherwise skip.](#please-read-if-you-are-using-googles-teachable-machine-otherwise-skip)
 * [How to add tflite model and label to flutter](#how-to-add-tflite-model-and-label-to-flutter)
 * [How to use this plugin](#how-to-use-this-plugin)
 * [Rough guide on parameters](#rough-guide-on-the-parameters)
 * [Android Installation & Permissions](#android-installation--permissions)
 * [iOS Installation & Permissions](#ios-installation--permissions)
 * [References](#references)
 
<br>

## About This Plugin

### The plugin has several features:

1. Audio recognition for stored audio files. (Only mono wav files for now)

2. Audio recognition for recordings. 

3. Tunable parameters for recording/inferences
   *  Please look a the [parameters](#rough-guide-on-the-parameters) below for more information.

4. Automatically reshape/transpose audio inputs.

<br>

### This plugin can support several model types:

1. Google Teachable Machine (Raw audio input)

   * For beginners with little to no machine learning knowledge. You can read can read the tutorial [here](https://carolinamalbuquerque.medium.com/audio-recognition-using-tensorflow-lite-in-flutter-application-8a4ad39964ae) if you are a newbie.
   * Training can be done [here](https://teachablemachine.withgoogle.com/train/audio) 

2. Raw audio input. 

   * Can recognize the following inputs: float32[audioLength, 1] or float32[1, audioLength]
   * For more information on how to train your own model, take a look [here](https://github.com/tensorflow/examples/tree/master/lite/examples/speech_commands/ml).

3. Decoded wav input. 

   * Supports two inputs: float32[audioLength, 1] and int32[1]
   * For more information on how to train your own model. Take a look [here](https://github.com/tensorflow/docs/blob/master/site/en/r1/tutorials/sequences/audio_recognition.md)
   * To train a decoded wave with MFCC, take a look [here](https://github.com/tensorflow/tensorflow/tree/r1.15/tensorflow/examples/speech_commands)

4. **(Experimental feature)**  Spectogram, melspectrogram, and MFCC inputs. 

   * Please note that this feature is experimental, and results may not be accurate compared to raw audio / decoded wav.
   * Spectrogram model can be trained here [tutorial](https://www.tensorflow.org/tutorials/audio/simple_audio). 

5. **(Currently worked on feature)** Multiple input and outputs.

<br>

## Known Issues/Commonly asked questions

1. **How to adjust the recording length/time**

   There are two ways to reduce adjust recording length/time:

   * You can increase the recording time by adjusting the bufferSize to a lower value. 
  
   * You can also increase recording time by lowering the sample rate.     
  
   **Note:** That stretching the value too low will cause problems with model accuracy. In that case, you may want to consider lowering your sample rate as well. Likewise, a very low sample rate can also cause problems with accuracy. It is your job to find the sweetspot for both values.

2. **How to reduce false positives in my model**

   To reduce false positives, you may want to adjust the default values of `detectionThreshold=0.3` and `averageWindowDuration=1000` to a higher value. A good value for both respectively are `0.7`  and `1500`. For more details about these parameters, please visit this [section](#rough-guide-on-the-parameters).

3. **I am getting build errors on iOS**

   There are several ways to fix this:

   * Some have reported to fix this issue by replacing the following line:

     ```ruby
     target 'Runner' do
       use_frameworks! 
       use_modular_headers!
       #pod 'TensorFlowLiteSelectTfOps' #Old line
       pod'TensorFlowLiteSelectTfOps','~> 2.6.0' #New line

       flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))
     end
     ```

   * Others have fixed this issue building the app without the line: `pod 'TensorFlowLiteSelectTfOps`. Then rebuilding the app by re-adding the line again.

   * Remember to run the following below:
   
   ```
     1. cd into iOS folder

     2. Run `flutter pub get` on terminal

     3. Run `pod install` on terminal

     4. Run `flutter clean` on terminal

     5. Run `flutter run` on terminal. All done!
   ```

4. **I am getting TensorFlow Lite Error on iOS. -  Regular TensorFlow ops are not supported by this interpreter. Make sure you apply/link the Flex delegate before inference** 

   *  Please make sure that you have enabled ops-select on your [podfile - step 4 & Xcode - step 5](#ios-if-you-are-using-googles-teachable-machine-model-otherwise-skip) and [build gradle - step 3](#android-if-you-are-using-googles-teachable-machine-otherwise-skip)

   * If you tried above, please run the example on a device (not emulator). If you still recieved this error, its very likely that theres an issue with cocoapod or Xcode configuration. Please check the [issue #7](https://github.com/Caldarie/flutter_tflite_audio/issues/7)

   * If you recieved this error from your custom model (not GTM), its likely that you're using unsupported tensorflow operators for tflite, as found in [issue #5](https://github.com/Caldarie/flutter_tflite_audio/issues/5#issuecomment-789260402). For more details on which operators are supported, look at the official documentation [here](https://www.tensorflow.org/lite/guide/ops_compatibility)

   * Take a looking at issue number 3 if none of the above works.  

5. **(iOS) App crashes when running Google's Teachable Machine model** 

   Please run your simulation on actual iOS device. Running your device on M1 macs should also be ok.
  
   As of this moment, there's [limited support](https://github.com/tensorflow/tensorflow/issues/44997#issuecomment-734001671) for x86_64 architectures from the Tensorflow Lite select-ops framework. If you absolutely need to run it on an emulator, you can consider building the select ops framework yourself. Instructions can be found [here](https://www.tensorflow.org/lite/guide/ops_select#ios)

6. **(Android) Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xfffffff4 in tid 5403** 

   It seems like the latest tflite package for android is causing this issue. Until this issue is fixed, please run this package on an actual Android Device. 

<br>

## Please Read If You Are Using Google's Teachable Machine. (Otherwise Skip)

<br>

**BE AWARE:** Google's Teachable Machine requires [select tensorflow operators](https://www.tensorflow.org/lite/guide/ops_select#using_bazel_xcode) to work. This feature is experimental and will cause the following issues:

1. Increase the overall size of your app. If this is unnacceptable for you, it's recommended that you build your own custom model. Tutorials can be found in the [About this plugin section](#about-this-plugin)

2. Emulators for iOS do not work due to limited support for x86_64 architectures. You need to run your simulation on an actual device. Issue can be found [here](https://github.com/tensorflow/tensorflow/issues/44997)
  
3. You will need to manually implement ops-select on your [podfile - step 4 & Xcode - step 5](#note-skip-below-if-your-are-not-using-google-teachable-machine-ios) and [build gradle - step 3](#note-skip-below-if-your-are-not-using-google-teachable-machine-android)

<br>

## How to add tflite model and label to flutter:

<br>

1. Create an assets folder and then place your custom tflite model and labels inside. 

![](https://github.com/Caldarie/flutter_tflite_audio/blob/master/pictures/model-label-asset.png)

2. In pubsec.yaml, link your tflite model and label under 'assets'. For example:

```
  assets:
    - assets/decoded_wav_model.tflite
    - assets/decoded_wav_label.txt

```

<br>

## How to use this plugin

<br>

Please look at the [example](https://github.com/Caldarie/flutter_tflite_audio/tree/master/example) on how to implement these futures.


1. To add the package in pubspec.yaml, open your terminal and run this line in your flutter project:

   ```
   flutter pub add tflite_audio
   ```

2. Import the plugin. For example:

   ```
   import 'package:tflite_audio/tflite_audio.dart';
   ```


3. To load your model:


   ```dart
   //Example for decodedWav models
    TfliteAudio.loadModel(
         model: 'assets/conv_actions_frozen.tflite',
         label: 'assets/conv_actions_label.txt',
         inputType: 'decodedWav');


   //Example for Google's Teachable Machine models
     TfliteAudio.loadModel(
         model: 'assets/google_teach_machine_model.tflite',
         label: 'assets/google_teach_machine_label.txt',
         inputType: 'rawAudio');

   //Example if you want to take advantage of all optional parameters from loadModel()
     TfliteAudio.loadModel(
       model: 'assets/conv_actions_frozen.tflite',
       label: 'assets/conv_actions_label.txt',
       inputType: 'decodedWav',
       outputRawScores: false, 
       numThreads: 1,
       isAsset: true,
     );
   ```


4. To start and listen to the stream for inference results:

    * Declare stream value
      ```dart
      Stream<Map<dynamic, dynamic>> recognitionStream;
      ```

    * If you want to use the recognition stream for recording:
      ```dart
      //Example values for Google's Teachable Machine models
      recognitionStream = TfliteAudio.startAudioRecognition(
        sampleRate: 44100,
        bufferSize: 22016,
        )

      //Example values for decodedWav
      recognitionStream = TfliteAudio.startAudioRecognition(
        sampleRate: 16000,
        bufferSize: 2000,
        )
        
      //Example for advanced users who want to utilise all optional parameters from this package. 
      //Note the values are default.
      recognitionStream = TfliteAudio.startAudioRecognition(
        sampleRate: 44100,
        bufferSize: 22016,
        numOfInferences: 5,
        audioLength = 44032,
        detectionThreshold: 0.3, 
        averageWindowDuration = 1000,
        minimumTimeBetweenSamples = 30,
        suppressionTime = 1500,
        )

      ```

    * If you want to use the recognition stream for stored audio files. 

       ```dart
       //Example values for Google teachable models
       recognitionStream = TfliteAudio.startFileRecognition(
         sampleRate: 44100,
         audioDirectory: "assets/sampleAudio.wav",
         );

      //Examples values for decodedWav
       recognitionStream = TfliteAudio.startFileRecognition(
         sampleRate: 16000,
         audioDirectory: "assets/sampleAudio.wav",
         );

       //Example for advanced users who want to utilise all optional parameters from this package. 
       recognitionStream = TfliteAudio.startFileRecognition(
         sampleRate: 44100,
         audioDirectory: "assets/sampleAudio.wav",
         audioLength: 44032,
         detectionThreshold: 0.3,
         averageWindowDuration: 1000,
         minimumTimeBetweenSamples: 30,
         suppressionTime: 1500,
         );
       ```

    * Listen for results 
      ```dart
      String result = '';
      int inferenceTime = 0;

      recognitionStream.listen((event){
            result = event["inferenceTime"];
            inferenceTime = event["recognitionResult"];
            })
          .onDone(
             //Do something here when stream closes
           );
      ```

5. To forcibly cancel recognition stream
  
   ```dart
   TfliteAudio.stopAudioRecognition();
   ```

</br>

## Rough guide on the parameters
  
  * outputRawScores - Will output the result as an array in string format. For example `'[0.2, 0.6, 0.1, 0.1]'`

  * numThreads -  Higher threads will reduce inferenceTime. However, will utilise the more cpu resource.

  * isAsset - is your model, label or audio file in the asset file? If yes, set true. If the files are outside (such as external storage), set false.
  
  * numOfInferences - determines how many times you want to loop the recording and inference. For example:
`numOfInference = 3` will repeat the recording three times, so recording length will be (1 to 2 seconds) x 3 = (3 to 6 seconds). Also the model will output the scores three times.

  * sampleRate - A higher sample rate may improve accuracy for recordings. Recommened values are 16000, 22050, 44100

  * audioLength - Default is 0 as the plugin will determine the length for you. You can manually adjust this if you wish to shorten or extend the number of audio samples.

  * bufferSize - A lower value will lengthen the recording. Likewise, a higehr value will shorten the recording. Make sure this value is equal or below your recording length. 
    
  * detectionThreshold - Will ignore any predictions where its probability does not exceed the detection threshold. Useful for situations where you pickup unwanted/unintentional sounds. Lower the value if your model's performance isn't doing too well.

  * suppressionMs - If your detection triggers too early, the result may be poor or inaccurate. Adjust the values to avoid this situation.

  * averageWindowDurationMs - Use to remove earlier results that are too old.

  * minimumTimeBetweenSamples - Ignore any results that are coming in too frequently

<br>

## Android Installation & Permissions

1. Add the permissions below to your AndroidManifest. This could be found in `<YourApp>/android/app/src`. For example:

   ```
   <uses-permission android:name="android.permission.RECORD_AUDIO" />
   <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
   ```

2. Edit the following below to your build.gradle. This could be found in `<YourApp>/app/src/`. For example:

   ```Gradle
   aaptOptions {
           noCompress 'tflite'
   ```

<br>

#### **NOTE:** Skip below if your are not using Google Teachable Machine (Android)

<br>


1. Enable select-ops under dependencies in your build gradle.

   ```Gradle
   dependencies {
       compile 'org.tensorflow:tensorflow-lite-select-tf-ops:+'
   }
   ```

<br>

## iOS Installation & Permissions

1. Add the following key to Info.plist for iOS. This could be found in `<YourApp>/ios/Runner`
   ```
   <key>NSMicrophoneUsageDescription</key>
   <string>Record audio for playback</string>
   ```

2. Change the deployment target to a minumum of 12.0 or higher. This could be done by:

   * Open your project workspace on xcode. Project workspace can be found here: `<YourApp>/ios/Runner.xcworkspace`
  
   * Select the top level Runner on the left panel
   
   * Select the Runner under Project.
  
   * Under the info tab, change the iOS deployment target to a minimum of 12.0 or higher

   ![](https://github.com/Caldarie/flutter_tflite_audio/blob/master/pictures/deployment-target.png)
    
3. Open your podfile (found here: `<YourApp>/ios/Podfile`) and change platform ios to a minimum 12 or higher. 

   ```ruby
   platform :ios, '12.0'
   ```

<br>

#### **NOTE:** Skip below if your are not using Google Teachable Machine (iOS)

<br>


1. In the same podfile, add `pod 'TensorFlowLiteSelectTfOps' under target.

   ```ruby
   target 'Runner' do
     use_frameworks! 
     use_modular_headers!
     pod'TensorFlowLiteSelectTfOps','~> 2.6.0' #Add this line here

     flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))
   end
   ```

2. Force load Select Ops for Tensorflow. To do that:

   * Open your project on xcode 
    
   * click on runner under "Targets"
  
   * Click on "Build settings" tab

   * Click on "All" tab

   * Click on the empty space which is on the right side of "Other Links Flag"

   * Add the following line: `-force_load $(SRCROOT)/Pods/TensorFlowLiteSelectTfOps/Frameworks/TensorFlowLiteSelectTfOps.framework/TensorFlowLiteSelectTfOps`

   ![](https://github.com/Caldarie/flutter_tflite_audio/blob/master/pictures/tflite-select-ops-installation.png)


3. Install the ops-select package to pod. To do this:

     * cd into iOS folder

     * Run `flutter pub get` on terminal

     * Run `pod install` on terminal

     * Run `flutter clean` on terminal

     * Run `flutter run` on terminal. All done!
    
<br>

## References

This project wouldn't of been possible if it wasn't for the following:

1. Project is based on:
   * https://github.com/tensorflow/examples/tree/master/lite/examples/speech_commands
2. Tflite & select ops:
   * https://www.tensorflow.org/lite/guide/ops_select
   * https://libraries.io/cocoapods/TensorFlowLiteSelectTfOps
3. Spectogram libraries:
   * https://github.com/Subtitle-Synchronizer/jlibrosa
   * https://github.com/dhrebeniuk/RosaKit
4. RxJava and RxSwift
   * https://github.com/ReactiveX/RxAndroid
   * https://github.com/ReactiveX/RxJava
   * https://github.com/ReactiveX/RxSwift

CHANGELOG.md
## 0.3.0
* BREAK CHANGE: Recording bufferSize now takes in 2x the number of samples. To keep the same recording length, simply divide your previous bufferSize by 2.
* Experimental: Support MFCC, melSpectrogram and spectrogram inputs
* Feature: Can automatically or manually set audio length
* Feature: Can automatically or manually transpose input shape
* Improvement: Stability of asyncronous operations with RxJava and RxSwift
* Improvement: (iOS) Removed meta info when extracting data from audio file.
* Improvement: (Android) Splicing algorithm passes all test case. Audio recogntion should now be more accurate.
* Fixed: (iOS) Duplicate symbol error. Set version of TensorFlowLite to 2.6.0. Problem found [here][i25].
* Fixed: (Android & iOS) Incorrect padding when splicing audio file. All test cases have passed.

[i25]: https://github.com/Caldarie/flutter_tflite_audio/issues/25
  
## 0.2.2+4
* Handled NaN exception for raw output on swift

## 0.2.2+3
* Hot fixed iOS issue where it will record indefintetly.

## 0.2.2+2
* Hot fixed missing AudioProcessing class.

## 0.2.2+1
* Hot fixed issue with unresponsive forced stop recognition.

## 0.2.2
* Feature: Added ability to recognise stored audio files
* Breaking Change: RecordingLength will no longer be required as a parameter.
* Fixed: NaN output for bufferRates that are non divisible to audioLength 
* Fixed: android permission error when granted outside app.

## 0.2.1+2
* Fixed NaN raw score output for Android.

## 0.2.1+1
* Fixed inaccurate numOfInference count for iOS and android.

## 0.2.1
* Improved recognition accuracy for Google Teachable Machine models
* Fixed memory crash on android
* Improved memory performance on iOS
* Added feature to output raw scores
* moved inputType to loadModel() instead of startAudioRecognition()

## 0.2.0
* Fixed crashed on Android when force stopping recognition
* Improve recognition latency on android by reducing number of event calls.

## 0.1.9
* Added support for android V2 embedding
* Breaking change - no longer supports deprecated versions of Android (pre 1.12)

## 0.1.8+2
* Fixed null safety incompatability with example 

## 0.1.8+1
* Fixed the problem with bridge NSNumber to Float
* Merged rawAudioRecognize() and decodedWavRecognize() on native platforms
* Set detection parameters to 0 for better performance.

## 0.1.8
* Added null safety compatability

## 0.1.7+1
* Hotfixed iOS crash when casting double to float for detectionThreshold

## 0.1.7
* Fixed iOS bug where stream wont close when permission has been denied.
* Added feature where you can adjust the detection sensitivity of the model

## 0.1.6+2
* Fixed podsec error
* Fixed iOS incompatability with fluter 2.0.0

## 0.1.6+1
* Hotfixed missing value for recording.

## 0.1.6
* bufferSize no longer needs to be divisible to recording length. 

## 0.1.5+3
* Fixed major android crash, where forcibly stopping the stream causes recorder.stop() to be prematurely called.
* Fixed minor iOS crash, where forcibly stopping the stream during recognition returns a nil exception.
* Cleaned up example for easy switch between decodedWav and Google's Teachable Machine model

## 0.1.5+2
* Disabled Google's Teachable Machine by default to reduce app footprint. (This can be enabled manually)
* Adjusted example's values to improve inference accuracy

## 0.1.5+1
* Added documentation
* Added example model from Google's Teachable Machine.
* Fixed iOS crash when loading text file with empty elements.

## 0.1.5
* Added support for Google Teachable Machine models.
* Fixed inaccurate reading with recording
* Added feature to switch between decodedwav and Google's Teachable machine model.

## 0.1.4
* Added a new feature where you can run multiple inferences per recording.
* Replaced future with stream when getting results from inferences
* Added a button to cancel the stream / inference
* Removed unnecessary code for easier reading.

## 0.1.3+1
* Used reusable widgets for easier to read code.
* Added some documentation 

## 0.1.3
* Hotfix for crash when permission has been denied.
* Added the key 'hasPermission' for the future startAudioRecognitions().
* Added feature in example where it'll show inference times

## 0.1.2
* Instead of returning a single string value, the future startAudioRecognition() now returns a map with the following keys:
    - recognitionResult 
    - inferenceTime
* Fixed issue in example where pressing the record button multiple times will crash the app.
* Added feature in example where pressing the recording button changes color.

## 0.1.1
* Made some fixes with making options explicit
* Added alert dialog when permission is denied.

## 0.1.0
* Added iOS support

## 0.0.4
* Added the following arguments into the future: startAudioRecognition()
    - sampleRate
    - audioLength
    - bufferSize

## 0.0.3
* Merged permission and audio recognition futures into one future.

## 0.0.2
* Fixed image url

## 0.0.1

* Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'dart:developer';
import 'package:tflite_audio/tflite_audio.dart';
import 'package:flutter/services.dart';
import 'dart:convert';

void main() => runApp(const MyApp());

///This example showcases how to take advantage of all the futures and streams
///from the plugin.
class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();
  final isRecording = ValueNotifier<bool>(false);
  Stream<Map<dynamic, dynamic>>? result;

  ///example values for decodedwav models
  // final String model = 'assets/decoded_wav_model.tflite';
  // final String label = 'assets/decoded_wav_label.txt';
  // final String audioDirectory = 'assets/sample_audio_16k_mono.wav';
  // final String inputType = 'decodedWav';
  // final int sampleRate = 16000;
  // final int bufferSize = 2000;
  // // final int audioLength = 16000;

  ///example values for google's teachable machine model
  final String model = 'assets/google_teach_machine_model.tflite';
  final String label = 'assets/google_teach_machine_label.txt';
  final String inputType = 'rawAudio';
  final String audioDirectory = 'assets/sample_audio_44k_mono.wav';
  final int sampleRate = 44100;
  final int bufferSize = 11016;
  // final int audioLength = 44032;

  ///example values for MFCC, melspectrogram, spectrogram models
  // final String model = 'assets/spectrogram_model.tflite';
  // final String label = 'assets/spectrogram_label.txt';
  // final String inputType = 'spectrogram';

  // final String model = 'assets/melspectrogram_model.tflite';
  // final String label = 'assets/melspectrogram_label.txt';
  // final String inputType = 'melSpectrogram';

  // final String model = 'assets/mfcc_model.tflite';
  // final String label = 'assets/mfcc_label.txt';
  // final String inputType = 'mfcc';

  // final String audioDirectory = 'assets/sample_audio_16k_mono.wav';
  // final int sampleRate = 16000;
  // final int bufferSize = 2000;
  // // final int audioLength = 16000;

  ///Optional parameters you can adjust to modify your input and output
  final bool outputRawScores = false;
  final int numOfInferences = 5;
  final int numThreads = 1;
  final bool isAsset = true;

  ///Adjust the values below when tuning model detection.
  final double detectionThreshold = 0.3;
  final int averageWindowDuration = 1000;
  final int minimumTimeBetweenSamples = 30;
  final int suppressionTime = 1500;

  @override
  void initState() {
    super.initState();
    TfliteAudio.loadModel(
      // numThreads: this.numThreads,
      // isAsset: this.isAsset,
      // outputRawScores: outputRawScores,
      inputType: inputType,
      model: model,
      label: label,
    );

    //spectrogram parameters
    // TfliteAudio.setSpectrogramParameters(nFFT: 256, hopLength: 129);

    // mfcc parameters
    TfliteAudio.setSpectrogramParameters(nMFCC: 40, hopLength: 16384);
  }

  void getResult() {
    ///example for stored audio file recognition
    // result = TfliteAudio.startFileRecognition(
    //   audioDirectory: audioDirectory,
    //   sampleRate: sampleRate,
    //   // audioLength: audioLength,
    //   // detectionThreshold: detectionThreshold,
    //   // averageWindowDuration: averageWindowDuration,
    //   // minimumTimeBetweenSamples: minimumTimeBetweenSamples,
    //   // suppressionTime: suppressionTime,
    // );

    ///example for recording recognition
    result = TfliteAudio.startAudioRecognition(
      sampleRate: sampleRate,
      bufferSize: bufferSize,
      numOfInferences: numOfInferences,
      // audioLength: audioLength,
      // detectionThreshold: detectionThreshold,
      // averageWindowDuration: averageWindowDuration,
      // minimumTimeBetweenSamples: minimumTimeBetweenSamples,
      // suppressionTime: suppressionTime,
    );

    ///Below returns a map of values. The keys are:
    ///"recognitionResult", "hasPermission", "inferenceTime"
    result
        ?.listen((event) =>
            log("Recognition Result: " + event["recognitionResult"].toString()))
        .onDone(() => isRecording.value = false);
  }

  ///fetches the labels from the text file in assets
  Future<List<String>> fetchLabelList() async {
    List<String> _labelList = [];
    await rootBundle.loadString(this.label).then((q) {
      for (String i in const LineSplitter().convert(q)) {
        _labelList.add(i);
      }
    });
    return _labelList;
  }

  ///handles null exception if snapshot is null.
  String showResult(AsyncSnapshot snapshot, String key) =>
      snapshot.hasData ? snapshot.data[key].toString() : '0 ';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
            key: _scaffoldKey,
            appBar: AppBar(
              title: const Text('Tflite-audio/speech'),
            ),

            ///Streambuilder for inference results
            body: StreamBuilder<Map<dynamic, dynamic>>(
                stream: result,
                builder: (BuildContext context,
                    AsyncSnapshot<Map<dynamic, dynamic>> inferenceSnapshot) {
                  ///futurebuilder for getting the label list
                  return FutureBuilder(
                      future: fetchLabelList(),
                      builder: (BuildContext context,
                          AsyncSnapshot<List<String>> labelSnapshot) {
                        switch (inferenceSnapshot.connectionState) {
                          case ConnectionState.none:
                            //Loads the asset file.
                            if (labelSnapshot.hasData) {
                              return labelListWidget(labelSnapshot.data);
                            } else {
                              return const CircularProgressIndicator();
                            }
                          case ConnectionState.waiting:

                            ///Widets will let the user know that its loading when waiting for results
                            return Stack(children: <Widget>[
                              Align(
                                  alignment: Alignment.bottomRight,
                                  child: inferenceTimeWidget('calculating..')),
                              labelListWidget(labelSnapshot.data),
                            ]);

                          ///Widgets will display the final results.
                          default:
                            return Stack(children: <Widget>[
                              Align(
                                  alignment: Alignment.bottomRight,
                                  child: inferenceTimeWidget(showResult(
                                          inferenceSnapshot, 'inferenceTime') +
                                      'ms')),
                              labelListWidget(
                                  labelSnapshot.data,
                                  showResult(
                                      inferenceSnapshot, 'recognitionResult'))
                            ]);
                        }
                      });
                }),
            floatingActionButtonLocation:
                FloatingActionButtonLocation.centerFloat,
            floatingActionButton: ValueListenableBuilder(
                valueListenable: isRecording,
                builder: (context, value, widget) {
                  if (value == false) {
                    return FloatingActionButton(
                      onPressed: () {
                        isRecording.value = true;
                        setState(() {
                          getResult();
                        });
                      },
                      backgroundColor: Colors.blue,
                      child: const Icon(Icons.mic),
                    );
                  } else {
                    return FloatingActionButton(
                      onPressed: () {
                        log('Audio Recognition Stopped');
                        TfliteAudio.stopAudioRecognition();
                      },
                      backgroundColor: Colors.red,
                      child: const Icon(Icons.adjust),
                    );
                  }
                })));
  }

  ///If snapshot data matches the label, it will change colour
  Widget labelListWidget(List<String>? labelList, [String? result]) {
    return Center(
        child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: labelList!.map((labels) {
              if (labels == result) {
                return Padding(
                    padding: const EdgeInsets.all(5.0),
                    child: Text(labels.toString(),
                        textAlign: TextAlign.center,
                        style: const TextStyle(
                          fontWeight: FontWeight.bold,
                          fontSize: 25,
                          color: Colors.green,
                        )));
              } else {
                return Padding(
                    padding: const EdgeInsets.all(5.0),
                    child: Text(labels.toString(),
                        textAlign: TextAlign.center,
                        style: const TextStyle(
                          fontWeight: FontWeight.bold,
                          color: Colors.black,
                        )));
              }
            }).toList()));
  }

  ///If the future isn't completed, shows 'calculating'. Else shows inference time.
  Widget inferenceTimeWidget(String result) {
    return Padding(
        padding: const EdgeInsets.all(20.0),
        child: Text(result,
            textAlign: TextAlign.center,
            style: const TextStyle(
              fontWeight: FontWeight.bold,
              fontSize: 20,
              color: Colors.black,
            )));
  }
}
